---
title: "Compositional analysis with Milo"
author: "Nadine Bestard"
date: "`r Sys.Date()`"
output: 
  html_document:
    code_folding: hide
  
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, 
                      dev = c('png', 'pdf'))
```

## Package

```{r message=FALSE, warning=FALSE}
library(miloR) # compositional analysis
library(here) # reproducible paths
library(scater) # sc plots
library(dplyr) # modify design df
```

This analysis have been done following the (MiloR vignette)[https://rawcdn.githack.com/MarioniLab/miloR/7c7f906b94a73e62e36e095ddb3e3567b414144e/vignettes/milo_gastrulation.html#5_Finding_markers_of_DA_populations]

# Load data

For this study we will load our processed single cell experiment 

```{r}
source(here("src/colours.R"))
project <- "fire-mice"
fig_path <- here("outs", project, "plots","DA_miloR", "/")
sce <- readRDS(here("processed", project, "sce_anno_02.RDS"))
```

# Visualize the data

```{r, dev="jpeg"}
plotReducedDim(sce, colour_by="genotype", dimred = "TSNE", text_by = "clusters_named") 
```

We will test for significant differences in abundance of cells between WT and KO, and the associated gene signatures.


# Differential abundance testing

## Create a Milo object

For differential abundance analysis on graph neighbourhoods we first construct a `Milo` object. This extends the [`SingleCellExperiment`](https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html) class to store information about neighbourhoods on the KNN graph. 

```{r}
milo <- Milo(sce)
milo
```


## Construct KNN graph

We need to add the KNN graph to the Milo object. This is stored in the `graph` slot, in [`igraph`](https://igraph.org/r/) format. The `miloR` package includes functionality to build and store the graph from the PCA dimensions stored in the `reducedDim` slot. 

For graph building you need to define a few parameters:

- `d`: the number of reduced dimensions to use for KNN refinement. We recommend using the same $d$ used for KNN graph building. In our case 26 dimensions (see feature_selection_dimred_02 script)
- `k`: this  affects the power of DA testing, since we need to have enough cells from each sample represented in a neighbourhood to estimate the variance between replicates. On the other side, increasing $k$ too much might lead to over-smoothing. We suggest to start by using the same value for $k$ used for KNN graph building for clustering and UMAP visualization. In our case k20.  We will later use some heuristics to evaluate whether the value of $k$ should be increased.

```{r}
# k modified after checking neighbourhoods
milo <- buildGraph(milo, k = 30, d = 25, reduced.dim = "PCA")
```

Alternatively, one can add a precomputed KNN graph (for example constructed with Seurat or scanpy) to the `graph` slot using the adjacency matrix, through the helper function `buildFromAdjacency`.

## Defining representative neighbourhoods on the KNN graph

We define the neighbourhood of a cell, the index, as the group of cells connected by an edge in the KNN graph to the index cell. For efficiency, we don't test for DA in the neighbourhood of every cell, but we sample as indices a subset of representative cells, using a KNN sampling algorithm used by [Gut et al. 2015](https://www.nature.com/articles/nmeth.3545). 

As well as $d$ and $k$, for sampling we need to define a few additional parameters:

- `prop`: the proportion of cells to randomly sample to start with. We suggest using `prop=0.1` for datasets of less than 30k cells. For bigger datasets using `prop=0.05` should be sufficient (and makes computation faster).
- `refined`: indicates whether you want to use the sampling refinement algorithm, or just pick cells at random. The default and recommended way to go is to use refinement. The only situation in which you might consider using `random` instead, is if you have batch corrected your data with a graph based correction algorithm, such as [BBKNN](https://github.com/Teichlab/bbknn), but the results of DA testing will be suboptimal.

```{r}
set.seed(1)
milo <- makeNhoods(milo, prop = 0.1, k = 30, d=25, refined = TRUE, reduced_dims = "PCA")
```

Once we have defined neighbourhoods, we plot the distribution of neighbourhood sizes (i.e. how many cells form each neighbourhood) to evaluate whether the value of $k$ used for graph building was appropriate. We can check this out using the `plotNhoodSizeHist` function. 

As a rule of thumb we want to have an average neighbourhood size over 5 x N_samples or to have a distribution peaking between 50 and 100. Otherwise you might consider rerunning makeNhoods increasing k and/or prop. In our case, 6 samples, an average of minimum 30 is expected, so we rerun makeNhood increasing k until we have an average of minimum 30 (5 x 6samples).

```{r}
plotNhoodSizeHist(milo)
```

## Counting cells in neighbourhoods

_Milo_ leverages the variation in cell numbers between replicates for the same experimental condition to test for differential abundance. Therefore we have to count how many cells from each sample are in each neighbourhood. We need to use the cell metadata and specify which column contains the sample information.

```{r}
milo <- countCells(milo, meta.data = as.data.frame(colData(milo)), sample="Sample")
```

This adds to the `Milo` object a $n \times m$ matrix, where $n$ is the number of neighbourhoods and $m$ is the number of experimental samples. Values indicate the number of cells from each sample counted in a neighbourhood. This count matrix will be used for DA testing.

```{r}
head(nhoodCounts(milo))
```

## Defining experimental design

Now we are all set to test for differential abundance in neighbourhoods. We implement this hypothesis testing in a generalized linear model (GLM) framework, specifically using the Negative Binomial GLM implementation in [`edgeR`](https://bioconductor.org/packages/release/bioc/html/edgeR.html).

We first need to think about our experimental design. The design matrix should match each sample to the experimental condition of interest for DA testing. In this case, we want to detect DA between genotypes, stored in the `genotype` column of the dataset `colData`. We also include the `chip` column in the design matrix. This represents a known technical covariate that we want to account for in DA testing. 

```{r}
design <- data.frame(colData(milo))[,c("Sample", "genotype", "chip")]
## Convert info from integers to factor
design$chip <- as.factor(design$chip) 
design$genotype <- as.factor(design$genotype)
design$genotype <- relevel(design$genotype, "WT")
# simplify data frame to only distinct combinations conditions
design <- distinct(design)
rownames(design) <- design$Sample
design
```



## Computing neighbourhood connectivity

Milo uses an adaptation of the Spatial FDR correction introduced by [cydar](https://bioconductor.org/packages/release/bioc/html/cydar.html), where we correct p-values accounting for the amount of overlap between neighbourhoods. Specifically, each hypothesis test P-value is weighted by the reciprocal of the kth nearest neighbour distance. To use this statistic we first need to store the distances between nearest neighbors in the Milo object. This is done by the `calcNhoodDistance` function
(N.B. this step is the most time consuming of the analysis workflow and might take a couple of minutes for large datasets).

```{r}
milo <- calcNhoodDistance(milo, d=25, reduced.dim = "PCA")
```


## Testing

Now we can do the DA test, explicitly defining our experimental design. In this case, we want to test for differences between genotype WT and KO, while accounting for the variability between technical batches (You can find more info on how to use formulas to define a testing design in R [here](https://r4ds.had.co.nz/model-basics.html#formulas-and-model-families))

```{r}
da_results <- testNhoods(milo, design = ~ chip + genotype, design.df = design)
head(da_results)
```

This calculates a Fold-change and corrected P-value for each neighbourhood, which indicates whether there is significant differential abundance between conditions. The main statistics we consider here are: 

- `logFC`: indicates the log-Fold change in cell numbers between samples from WT and KO
- `PValue`: reports P-values before FDR correction
- `SpatialFDR`: reports P-values corrected for multiple testing accounting for overlap between neighbourhoods

```{r}
da_results %>%
  arrange(SpatialFDR) %>%
  head() 
```

# Inspecting DA testing results

We can start inspecting the results of our DA analysis from a couple of standard diagnostic plots. 
We first inspect the distribution of uncorrected P values, to verify that the test was balanced.

```{r, p-value, fig.path=fig_path}
ggplot(da_results, aes(PValue)) + geom_histogram(bins=50)
```

Then we visualize the test results with a volcano plot (remember that each point here represents a neighbourhood, _not_ a cell).

```{r spatialFDR, fig.path=fig_path}
ggplot(da_results, aes(logFC, -log10(SpatialFDR))) + 
  geom_point() +
  geom_hline(yintercept = 1) ## Mark significance threshold (10% FDR)
```

The neighbourhoods with strong down regulation are the microglia. 

To visualize DA results relating them to the embedding of single cells, we can build an abstracted graph of neighbourhoods that we can superimpose on the single-cell embedding. Here each node represents a neighbourhood, while edges indicate how many cells two neighbourhoods have in common. Here the layout of nodes is determined by the position of the index cell in the UMAP embedding of all single-cells. The neighbourhoods displaying significant DA are coloured by their log-Fold Change.

```{r nhoodGraph, fig.width=15, fig.height=8, fig.path=fig_path}
milo <- buildNhoodGraph(milo)
## Plot single-cell TSNE
tsne_pl <- plotReducedDim(milo, dimred = "TSNE", colour_by="genotype", text_by = "clusters_named", 
                          text_size = 3, point_size=0.5) + 
  guides(fill="none")  + scale_color_manual(values = c(col_wt_ko[1],col_wt_ko[2])) + labs(color="genotype")

## Plot neighbourhood graph
nh_graph_pl <- plotNhoodGraphDA(milo, da_results, layout="TSNE",alpha=0.1) +  scale_fill_gradient2(high = scales::muted("red"), mid = "white", low = scales::muted("blue")) + labs(fill = "logFC")
  
tsne_pl + nh_graph_pl #+
 #plot_layout(guides="collect")
```


We might also be interested in visualizing wheather DA is particularly evident in certain clusters. To do this, we assign a cluster label to each neighbourhood by finding the most abundant cluster within cells in each neighbourhood. We can label neighbourhoods in the results `data.frame` using the function `annotateNhoods`. This also saves the fraction of cells harbouring the label.

```{r}
da_results <- annotateNhoods(milo, da_results, coldata_col = "clusters_named")
da_results <- annotateNhoods(milo, da_results, coldata_col = "celltype")
head(da_results)
```

While neighbourhoods tend to be homogeneous, we can define a threshold for `celltype_fraction` to exclude neighbourhoods that are a mix of cell types. 

```{r}
ggplot(da_results, aes(clusters_named_fraction)) + geom_histogram(bins=50)
```
```{r}
da_results$celltype <- ifelse(da_results$celltype_fraction < 0.7, "Mixed", da_results$celltype)
da_results$clusters_named <- ifelse(da_results$clusters_named_fraction < 0.7, "Mixed", da_results$clusters_named)
```

Now we can visualize the distribution of DA Fold Changes in different cell types or clusters

```{r DAbeeswarm, fig.path=fig_path}
# reorder factor
da_results$clusters_named <- factor(da_results$clusters_named, levels =c("Mixed", rev(levels(sce$clusters_named))))

plotDAbeeswarm(da_results, group.by = "clusters_named") +
  scale_colour_gradient2(high = scales::muted("red"), mid = "white", low = scales::muted("blue")) + xlab("")

```

```{r}
#save results
saveRDS(da_results, here("processed", project, "da_results.rds"))
```



<details>
  <summary>**Session Info**</summary>
  
```{r}
sessionInfo()
```

</details>